### ISSUE TITLE
Confusion between four rules wrt what constitutes a code block

### ISSUE THREAD

**Issue by srenfo** [markdownlint/markdownlint#318]
## Description

Rules MD029, MD031, MD032 and MD040 appear to have slightly different definitions of what constitutes a code block. This issue may supersede #299.

### Environment

Debian buster
**MDL Version** 0.9.0

### Expected Behavior

Code block handling should be consistent.

### Actual Behavior

Depending on whether or not code blocks are directly adjacent to a list, different rules are correctly or incorrectly triggered.

## Replication Case

"Compact" case:

~~~
# foobar

- foobar
  ```
  1. foo
  2. bar
  ```
~~~
```sh
$ mdl -s all foobar.md 
foobar.md:6: MD029 Ordered list item prefix
foobar.md:4: MD031 Fenced code blocks should be surrounded by blank lines
```

Dedented code block:

~~~
# foobar

- foobar
```
1. foo
2. bar
```
~~~
```sh
$ mdl -s all foobar.md 
foobar.md:4: MD031 Fenced code blocks should be surrounded by blank lines
foobar.md:3: MD032 Lists should be surrounded by blank lines
```
With blank line; code block indent irrelevant:
~~~
# foobar

- foobar

  ```
  1. foo
  2. bar
  ```
~~~
```sh
$ mdl -s all foobar.md 
foobar.md:5: MD040 Fenced code blocks should have a language specified
```

:exclamation:  NB. I assume a properly indented code block is a continuation of the list item and would be rendered as such. If that is not the case then the following interpretation is invalid.

Apparently MD029, MD031, MD032 and MD040 use different logic to find code blocks. I'm pretty sure only the last example ("with blank line") is handled correctly by all four rules, presumably because the blank line makes it very clear to the parser that list and code block are separate entities.

- MD029 doesn't classify the code block as such in the "compact" case but does so in the other two.
- MD031 correctly identifies all code blocks. It does not allow a code block to be a continuation of a list item (by triggering in the "compact" case), in contradiction to MD032.
- MD032 correctly identifies all code blocks. It **does** allow a code block to continue a list item (by not triggering in the "compact" case), in contradiction to MD031.
- MD040 doesn't find the code block in either of the first two cases. It should trigger in all three cases instead.

**Comment by jaymzh** [comment]
Thanks @srenfo ! This is super useful. Next time I have some time to sit down and hack on MDL I'll see if we can factor out the code here so that there's not variances in it.
